File: C#Stack/Stack/Stack.cs
using System;
using System.Runtime.CompilerServices;
using ConditionalWeakTableExtensions;

namespace Stack {
    public class Stack<T> 
        where T : class
    {
        private delegate R MatchDelegate<R>(Func<R> onEmpty, Func<T, Stack<T>, R> onElem);
        private Stack() {}
        private ConditionalWeakTable<T, Stack<T>> cache = new ConditionalWeakTable<T, Stack<T>>();
        private R Match<R>(MatchDelegate<R> matchDelegate, Func<R> onEmpty, Func<T, Stack<T>, R> onElem) {
            matchDelegate(onEmpty, onElem);
        }
        // public virtual R Match<R>(Func<R> onEmpty, Func<T, Stack<T>, R> onElem) {
        //     //MatchDelegate<R> del = new MatchDelegate<R>(onEmpty);
        // }
        public Stack<T> Push(T elem){return cache.GetValueOrCompute(elem, (elem) => this.NewPush(elem));}
        private Stack<T> NewPush(T elem) {
            Stack<T> self = this;
            self.Match<R>(Empty(), (e,t) = t.Push(e));
            return self;
        }
        private static Stack<T> _empty = new Stack<T>();
        public static Stack<T> Empty() {return _empty;}
        public R Match<R>(Func<R> onEmpty, Func<T,R> onLast, Func<T, Stack<T>, R> onElem) {
            return Match(onEmpty, (e,t) => t.IsEmpty() ? onLast(e):onElem(e,t));
        }
        private string ToStrAux(){return Match(() =>"]",e=>e+"]",(e,t)=>e+"; " + t.ToStrAux());}
        public override string ToString() {return "["+ToStrAux();}
        public bool IsEmpty() {return Match(() => true, (e,t) => false);}
        public T TopOrElse(Func<T> s){return Match(s, (e,t)=>e);}
        public Stack<T> PopOrElse(Func<Stack<T>> s){return Match(s, (e,t)=>t);}
    }
}

File: Stack/src/Stack.java
import java.util.WeakHashMap;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Supplier;

public class Stack<T>{
    private Stack(){} //Private constructor
    private WeakHashMap<T, Stack<T>> cache = new WeakHashMap<>();//one for each stack object
    public <R> R match(Supplier<R> onEmpty, BiFunction<T,Stack<T>,R> onElem){
        return onEmpty.get();
    }
    public Stack<T> push(T elem){return cache.computeIfAbsent(elem,this::newPush);}
    private Stack<T> newPush(T elem){//same implementation as before, but in a private method
        Stack<T> self=this;//explicit this naming
        return new Stack<T>(){//closure
            public <R> R match(Supplier<R> onEmpty, BiFunction<T,Stack<T>,R> onElem){
                return onElem.apply(elem,self);
            }};}
    private static Stack<Object> empty=new Stack<Object>();//singleton pattern
    @SuppressWarnings("unchecked")public static <T> Stack<T> empty(){return (Stack<T>)empty;}
    public <R> R match(Supplier<R> onEmpty, Function<T,R> onLast, BiFunction<T,Stack<T>,R> onElem){
        return match(onEmpty,(e,t)->t.isEmpty()?onLast.apply(e):onElem.apply(e,t));
    }
    private String toStrAux(){return match(()->"]",e->e+"]",(e,t)->e+"; "+t.toStrAux());}
    public String toString(){return "["+toStrAux();}
    public boolean isEmpty(){return match(()->true,(e,t)->false);}
    public T topOrElse(Supplier<T> s){return match(s,(e, t)->e);}
    public Stack<T> popOrElse(Supplier<Stack<T>> s){return match(s,(e,t)->t);}
    //note: no need of equals/hashcode any more!! The identity is now the right one!
}
